src/utils.cpp

```diff
@@ -521,13 +566,70 @@ std::vector<std::string> resolve_binary_path(const std::string& cmd)
     candidate_paths = ::expand_wildcard_paths(candidate_paths);
 
   std::vector<std::string> valid_executable_paths;
-  for (const auto& path : candidate_paths)
-    if (access(path.c_str(), X_OK) == 0)
-      valid_executable_paths.push_back(path);
+  for (const auto &path : candidate_paths) {
+    std::string rel_path;
+    if (pid > 0 && pid_in_different_mountns(pid))
+      rel_path = path_for_pid_mountns(pid, path);
+    else
+      rel_path = path;
+    if (access(rel_path.c_str(), X_OK) == 0)
+      valid_executable_paths.push_back(rel_path);
+  }
 
   return valid_executable_paths;
 }
 
+std::string path_for_pid_mountns(int pid, const std::string &path) {
+  std::ostringstream pid_relative_path;
+  std::string root =
+      (path.length() >= 1 && path.at(0) == '/') ? "/root" : "/root/";
+  pid_relative_path << "/proc/" << pid << root << path;
+  return pid_relative_path.str();
+}
+
+/*
+Determines if the target process is in a different mount namespace from
+bpftrace.
+
+If a process is in a different mount namespace (eg, container) it is very
+likely that any references to local paths will not be valid, and that paths
+need to be made relative to the PID.
+
+If an invalid PID is specified or doesn't exist, it returns false.
+True is only returned if the namespace of the target process could be read and
+it doesn't match that of bpftrace.
+*/
+static bool pid_in_different_mountns(int pid) {
+
+  const size_t BUFSIZE = 64;
+  char my_ns[BUFSIZE];
+  char pid_ns[BUFSIZE];
+  char pid_path[BUFSIZE];
+  size_t bytes;
+
+  if (pid <= 0)
+    return false;
+
+  bytes = readlink("/proc/self/ns/mnt", my_ns, BUFSIZE - 1);
+  if (bytes > 0 && bytes < BUFSIZE)
+    my_ns[bytes] = '\0';
+  else
+    return false;
+
+  snprintf(pid_path, BUFSIZE, "/proc/%d/ns/mnt", pid);
+  bytes = readlink(pid_path, pid_ns, BUFSIZE - 1);
+
+  if (errno == EACCES || errno == ENOENT || bytes > BUFSIZE) {
+    std::cerr << "Error reading mount ns from " << pid_path;
+    std::cerr << " for pid " << pid << std::endl;
```

Comment:
can you add the error here too? `strerror(errno)` iirc

Reply:
Done, I noticed this is failing on CI, I wonder if it will show a permission error on the next run...

Reply:
I figured out the issue, comment below.